cmake/test/gtest.cmake

```diff
@@ -287,118 +291,131 @@ function(catkin_find_google_test_source include_path src_path
-                                     gtest_include_dir gtest_lib_dir gtest_libs gtest_main_libs
-                                     gmock_found gmock_include_dir gmock_lib_dir gmock_libs
-                                     gmock_main_libs base_dir)
-      if(gtest_found)
-        set(GTEST_FROM_SOURCE_FOUND ${gtest_found} CACHE INTERNAL "")
-        set(GTEST_FROM_SOURCE_INCLUDE_DIRS ${gtest_include_dir} CACHE INTERNAL "")
-        set(GTEST_FROM_SOURCE_LIBRARY_DIRS ${gtest_lib_dir} CACHE INTERNAL "")
-        set(GTEST_FROM_SOURCE_LIBRARIES ${gtest_libs} CACHE INTERNAL "")
-        set(GTEST_FROM_SOURCE_MAIN_LIBRARIES ${gtest_main_libs} CACHE INTERNAL "")
-        if(gmock_found)
-          set(GMOCK_FROM_SOURCE_FOUND ${gmock_found} CACHE INTERNAL "")
-          set(GMOCK_FROM_SOURCE_INCLUDE_DIRS ${gmock_include_dir} CACHE INTERNAL "")
-          set(GMOCK_FROM_SOURCE_LIBRARY_DIRS ${gmock_lib_dir} CACHE INTERNAL "")
-          set(GMOCK_FROM_SOURCE_LIBRARIES ${gmock_libs} CACHE INTERNAL "")
-          set(GMOCK_FROM_SOURCE_MAIN_LIBRARIES ${gmock_main_libs} CACHE INTERNAL "")
-          message(STATUS "Found gmock sources under '${base_dir}': gmock will be built")
-        endif()
-        if(base_dir)
-          # overwrite CMake install command to skip install rules for gtest targets
-          # which have been added in version 1.8.0
-          _use_custom_install()
-          set(_CATKIN_SKIP_INSTALL_RULES TRUE)
-          add_subdirectory(${base_dir} ${gtest_lib_dir})
-          set(_CATKIN_SKIP_INSTALL_RULES FALSE)
-          set_target_properties(${gtest_libs} ${gtest_main_libs}
-                                PROPERTIES EXCLUDE_FROM_ALL 1)
-          if(gmock_found)
-            set_target_properties(${gmock_libs} ${gmock_main_libs}
-                                  PROPERTIES EXCLUDE_FROM_ALL 1)
-          endif()
-        endif()
-        message(STATUS "Found gtest sources under '${base_dir}': gtests will be built")
-      else()
-        if(CATKIN_TOPLEVEL)
-          message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system or checkout gtest (by running 'git clone  https://github.com/google/googletest.git -b release-1.8.0' in the source space '${CMAKE_SOURCE_DIR}' of your workspace) to enable gtests")
-        else()
-          message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system to enable gtests")
-        endif()
+find_package(GTest QUIET)
+if(NOT GMOCK_FOUND OR NOT GTEST_FOUND)
+  # If we find one but not the other, see if we can get both from source
+  # only add gmock/gtest directory once per workspace
+  if(NOT TARGET gtest AND NOT TARGET gmock)
+    # Fall back to system-installed gmock source (e.g. Ubuntu)
+    set(_include_paths "/usr/include" "/usr/src")
+    set(_source_paths "/usr/src")
+
+    catkin_find_google_test_source("${_include_paths}" "${_source_paths}" gtest_found
```

Comment:
@sloretz As pointed out in https://github.com/ros/catkin/issues/1030#issuecomment-540263776 passing a list of include paths to `catkin_find_google_test_source` is wrong here. The function only expects a single path  - e.g. it appends a subdirectory to it (`set(_gtest_include_paths "${include_path}/gtest")`). So either only a single path should be passed or the function needs to be updated to append the subdirectory to each item of the list. I am not sure which of the two options is the right way of resolving this.

Reply:
> the function needs to be updated to append the subdirectory to each item of the list or

I think this would be the correct fix

Reply:
Sounds good to me. Please create a PR with the change.

Reply:
See #1040.